File: DSAlgorithms/Leetcode/c++/232-implementQueueUsingStack.cpp
#include <iostream>
#include <stack>
class Queue {
    private:
        std::stack<int> stack_in;
        std::stack<int> stack_out;
    public:
        Queue(){};
        void push(int x){
            stack_in.push(x);
        }
        int pop()
        {
            peek();
            int val = stack_out.top();
            stack_out.pop();
            return val;
        }

        int peek()
        {
            if(stack_out.empty())
                while(!stack_in.empty())
                {
                    stack_out.push(stack_in.top());
                    stack_in.pop();
                }
            return stack_out.top();
        }

        bool empty()
        {
            return stack_in.empty() && stack_out.empty();
        }
        
};

File: dsa_algo/leetcode/ts/src/old/232-implementQueueUsingStack.ts
class Queue{
       stack_in: number[] = [];
       stack_out: number[] = [];
    constructor() {}
    push(x: number): void {
        this.stack_in.push(x);
    }
    pop(): number {
        this.peek();
        const valueToPop = this.stack_out[this.stack_out.length - 1];
        this.stack_out.pop();
        return valueToPop;
    }
    peek(): number {
        if(this.stack_out.length === 0)
            while(this.stack_in.length !==0)
            {
                this.stack_out.push(this.stack_in[this.stack_in.length - 1]);
                this.stack_in.pop();
            }
            return this.stack_out[this.stack_out.length - 1];
    }
    empty(): boolean {
        return (this.stack_in.length === 0 && this.stack_out.length === 0)
    }
}